src/run.rs

```diff
@@ -0,0 +1,171 @@
+                }
+            }
+            Err(e) => {
+                if flag_print {
+                    println!("{}: {}", file_name, e);
+                }
+                errors += 1;
+            }
+        }
+    }
+
+    if flag_print {
+        match total {
+            0 => println!("0 files"),
+            1 => println!("1 file"),
+            n => println!("{} files", n),
+        }
+    }
+
+    match errors {
+        0 => Ok(()),
+        1 => Err(String::from("1 failure")),
+        n => Err(format!("{} failures", n)),
+    }
+}
+
+/// Build an ISA based on the current machine running this code (the host)
+fn create_target_isa() -> Box<dyn TargetIsa> {
+    let settings_builder = settings::builder();
+    // b.set("opt_level", "fastest");
+    let flags = settings::Flags::new(settings_builder);
+
+    // build an ISA for the current machine
+    host_isa_builder()
+        .expect("A builder could not be created")
+        .finish(flags)
+}
+
+/// Iterate over all of the files passed as arguments. TODO eventually this should recurse into
+/// subdirectories, ignore non-CLIF files, ignore hidden files, etc.
+fn iterate_files(files: Vec<String>) -> Box<dyn Iterator<Item = PathBuf>> {
+    let iterator = files.into_iter().map(PathBuf::from);
+    Box::new(iterator)
+}
+
+/// Determine a path's file name
+fn get_file_name(path: &PathBuf) -> Result<String, String> {
+    let os_str = path
+        .file_name()
+        .ok_or(String::from("Unable to figure out the file name"))?;
+    Ok(String::from(os_str.to_string_lossy()))
+}
+
+/// Run all functions in a file that are succeeded by "run:" comments
+fn run_single_file(path: &PathBuf, isa: &dyn TargetIsa) -> Result<(), String> {
+    let file_contents = read_to_string(&path).map_err(|e| e.to_string())?;
+    run_file_contents(file_contents, isa)
+}
+
+/// Main body of `run_single_file` separated for testing
+fn run_file_contents(file_contents: String, isa: &dyn TargetIsa) -> Result<(), String> {
+    let test_file = parse_test(&file_contents, None, None).map_err(|e| e.to_string())?;
+    for (func, Details { comments, .. }) in test_file.functions {
+        // TODO this may need to iterate over the comments
+        if comments.iter().filter(|c| c.text.contains("run:")).count() > 0 {
+            run_single_function(func, isa)?
+        }
+    }
+    Ok(())
+}
+
+/// Compile and execute a single function, expecting a boolean to be returned; a 'true' value is
+/// interpreted as a successful test execution and mapped to Ok whereas a 'false' value is
+/// interpreted as a failed test and mapped to Err.
+fn run_single_function(func: Function, isa: &dyn TargetIsa) -> Result<(), String> {
+    if !(func.signature.params.is_empty()
+        && func.signature.returns.len() == 1
+        && func.signature.returns.first().unwrap().value_type.is_bool())
+    {
+        return Err(String::from(
+            "Functions must have a signature like: () -> boolean",
```

Comment:
Maybe allow`() -> ()` too, for tests which should not crash.

Reply:
It's been discussed outside the code, and it's a good design that's overall more global than what () -> () allows, so that looks fine to me.